### DISCUSSION TITLE
[JDA Gateway-Worker 1] WARN WebSocketClient - Missed 2 heartbeats! Trying to reconnect...

### DISCUSSION THREAD

**Discussion by Yellowstrawberrys** [discord-jda/JDA#1735] [Questions and Help] [answered]
When I start the bot, then in log "[JDA Gateway-Worker 1] WARN WebSocketClient - Missed 2 heartbeats! Trying to reconnect..." this message comes out but my internet is not bad at all(30mbps / 10mbps) and I googled it how to fix it but all answer was internet is bad; please help me :( 
I restarted too.

I'm using raspberry pi 4b 8Gb for hosting bot
and Germany internet

**Comment by MinnDevelopment** [comment]
Converting to discussion, since you didn't fill out any of the issue templates. See #1337

**Comment by DManstrator** [comment, marked as answer]
This means your listener methods take too long to finish. Therefore other events will be delayed and this warning comes up. This can have two reasons:

1. You use blocking operations such as `complete()` or `submit().get()` which will block the current code until the action was done and therefor can take very long.
2. You have expensive operations such as image manipulation or API calls which can take (very) long.

If it's the first case you should use `queue()` with the success consumer instead, where the consumer will be asynchronously executed when the action (such as sending messages, retrieving users, etc.) was done. Read more in the [wiki](https://github.com/DV8FromTheWorld/JDA/wiki/7%29-Using-RestAction) about it.

If it's the second case, you should process such commands on a new Thread. For this you should use a [`ScheduledExecutorService`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html). Check the linked docs to see how to create the ScheduledExecutorService but make sure you only create it once. In the end you would want to execute the command with `execute()`, `submit()` or `schedule()`, whatever fits you the most.

**Reply by Yellowstrawberrys** [reply]
Hi! Thank you for answering my question! I have question for second case... you said that I need to process commands on a new Thread and I need to use ScheduledExecutorService but can I do like that `Thread t = new Thread(() -> { //image upload });`  for uploading image?

**Reply by DManstrator** [reply]
That would create a new Thread every time you upload an image which is bad. Thus you should use a `ScheduledExecutorService` which manages a fixed amount of threads for you to do stuff on those threads.

PS: If you upload those images to Discord, you can use the success consumer of `sendFile(...).queue()`.

**Reply by Yellowstrawberrys** [reply]
Thank you! I fix my bot and it doesn't turn off :)
